### PR TITLE
Add rpi realhw usb keyboard emulation option

### PR DESCRIPTION
**Pull request by asdil12**
Ticket: https://progress.opensuse.org/issues/119566

image

### CODE REVIEW

**Review by okurz — CHANGES_REQUESTED** [review]
If you really want to distribute such code in os-autoinst-distri-opensuse, fine, though I think including that from elsewhere is a better idea. For new files please also add a copyright and license file, same as in test modules

The git commit message subject line should not end in a dot

File: data/generalhw_scripts/rpi_pico_w_keyboard/rpi_pico_w_keyboard.ino [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+
+const char *ssid = STASSID;
+const char *password = STAPSK;
+WebServer server(80);
+
+
+// Ringbuffer for the last 10 keyboard inputs to be shown on the webui
+#define LOGBUF_LEN 10
+char* logbuf[LOGBUF_LEN] = {NULL};
+uint8_t logbuf_next_idx = 0;
```
Comment (okurz) [inline data/generalhw_scripts/rpi_pico_w_keyboard/rpi_pico_w_keyboard.ino:22]: I recommend to use `uint_fast8_t` unless you need exactly 8 bit. Same recommendation goes for other uses of bit-fixed types.

File: data/generalhw_scripts/rpi_pico_w_keyboard/rpi_pico_w_keyboard.ino [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+
+void logbuf_add(const char* logline) {
+  if (logbuf[logbuf_next_idx] != NULL) {
+    free(logbuf[logbuf_next_idx]);
+  }
+  //Serial.println("LOG:");
+  //Serial.println(logline);
```
Comment (okurz) [inline data/generalhw_scripts/rpi_pico_w_keyboard/rpi_pico_w_keyboard.ino:29]: Don't add disabled code as nobody but you will understand why this is not active. Better leave it out and keep in your local git repository, either in `git stash` or a temporary "WIP"-commit

See https://github.com/os-autoinst/os-autoinst-distri-opensuse/blob/master/CONTRIBUTING.md#coding-style for more details

File: data/generalhw_scripts/rpi_pico_w_keyboard/rpi_pico_w_keyboard.ino [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+    uint8_t key_raw = get_key(key);
+    if (key_raw) { 
+      Serial.println("Pressing " + key);
+      Keyboard.press(key_raw);
+    }
+    else {
+      Serial.println("Could not find key " + key);
+    }
+    start = next + 1;
+    delay(30);
```
Comment (okurz) [inline data/generalhw_scripts/rpi_pico_w_keyboard/rpi_pico_w_keyboard.ino:60]: magic number, at least provide a named constant

File: data/generalhw_scripts/rpi_pico_w_keyboard/rpi_pico_w_keyboard.ino [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+
+  Serial.begin(115200);
+  WiFi.mode(WIFI_STA);
+  WiFi.begin(ssid, password);
+  Serial.println("");
+
+  // Wait for connection
+  uint16_t i = 0;
+  while (WiFi.status() != WL_CONNECTED) {
+    delay(500);
```
Comment (okurz) [inline data/generalhw_scripts/rpi_pico_w_keyboard/rpi_pico_w_keyboard.ino:85]: magic number

File: data/generalhw_scripts/rpi_pico_w_keyboard/rpi_pico_w_keyboard.ino [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+    Serial.print(".");
+    if (i++ > 30) {
```
Comment (okurz) [inline data/generalhw_scripts/rpi_pico_w_keyboard/rpi_pico_w_keyboard.ino:87]: magic number

File: data/generalhw_scripts/rpi_pico_w_keyboard/rpi_pico_w_keyboard.ino [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+}
+
+void loop(void) {
+  server.handleClient();
+}
+
+
+// Webserver callback functions
+void web_handle_root() {
+  char temp[8000];
```
Comment (okurz) [inline data/generalhw_scripts/rpi_pico_w_keyboard/rpi_pico_w_keyboard.ino:116]: magic number

File: data/generalhw_scripts/rpi_pico_w_keyboard/rpi_pico_w_keyboard.ino [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+  int hr = min / 60;
+
+  String ip = WiFi.localIP().toString();
+  String mac = WiFi.macAddress();
+
+  String log;
+  log.reserve(7000);
+  logbuf_out(&log);
+
+  snprintf(temp, 7000, "<html>\
```
Comment (okurz) [inline data/generalhw_scripts/rpi_pico_w_keyboard/rpi_pico_w_keyboard.ino:128]: magic number

File: data/generalhw_scripts/rpi_pico_w_keyboard/rpi_pico_w_keyboard.ino [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+  }
+
+  server.send(404, "text/plain", message);
+  digitalWrite(LED_BUILTIN, HIGH);
+}
+
+void web_handle_cmd() {
+  digitalWrite(LED_BUILTIN, LOW);
+  String out;
+  out.reserve(2600);
```
Comment (okurz) [inline data/generalhw_scripts/rpi_pico_w_keyboard/rpi_pico_w_keyboard.ino:181]: magic number

**Review by okurz — CHANGES_REQUESTED** [review]

File: data/generalhw_scripts/rpi_pico_w_keyboard/keymap.h [hunk 9]
```diff
@@ -0,0 +1,98 @@
+/*
+# Copyright 2022 SUSE LLC
```
Comment (okurz) [inline data/generalhw_scripts/rpi_pico_w_keyboard/keymap.h:2]: Please avoid copyright years in newly added files

```suggestion
# Copyright SUSE LLC
```

File: data/generalhw_scripts/rpi_pico_w_keyboard/rpi_pico_w_keyboard.ino [hunk 10]
```diff
@@ -0,0 +1,207 @@
+/*
+# Copyright 2022 SUSE LLC
```
Comment (okurz) [inline data/generalhw_scripts/rpi_pico_w_keyboard/rpi_pico_w_keyboard.ino:2]: same as above

File: data/generalhw_scripts/rpi_pico_w_keyboard/rpi_pico_w_keyboard.ino [hunk 11]
```diff
@@ -0,0 +1,207 @@
...
+#define WIFI_CONNECT_RESET_AFTER_INTERVAL_NUM 30
+
+#define DEFAULT_WEB_DOC_LEN 2000
+#define LOGBUF_LEN 10
+#define LOGBUF_MAXLEN DEFAULT_WEB_DOC_LEN * LOGBUF_LEN
+#define WEBROOT_LEN LOGBUF_MAXLEN + DEFAULT_WEB_DOC_LEN
+
+const char *ssid = STASSID;
+const char *password = STAPSK;
+WebServer server(80);
```
Comment (okurz) [inline data/generalhw_scripts/rpi_pico_w_keyboard/rpi_pico_w_keyboard.ino:28]: magic number

**Review by okurz — CHANGES_REQUESTED** [review]

**Review by okurz — APPROVED** [review]

**Review by ggardet — APPROVED** [review]
Awesome! I do not have RPi Pico W yet, so I cannot test, but I guess you did.

**Pull request merged**